### PR TITLE
Adds a graph concept to the kustomization detail page

### PR DESCRIPTION
**Pull request by jpellizzari**
Adds an initial graph concept. This will need to be re-worked and styled in the future.

Screenshot from 2021-03-18 11-16-19

### CODE REVIEW

**Review by bia — APPROVED** [review]
LGTM!
Looking forward to styling this and deep linking each node in the graph to its appropriate page.

Sometimes the graph loads off-centre, it gets fixed if I reload the page : 
Screenshot 2021-03-19 at 00 07 08

**Comment by bia** [comment]
..and why not add the full graph to each source detail page?

**Review by stefanprodan — CHANGES_REQUESTED** [review]
A dependency graph should list a tree based on Kustomization.spec.dependsOn. Can we please change the name to something else?

**Comment by jpellizzari** [comment]
Change from "Dependency Graph" to "Graph"

**Comment by jpellizzari** [comment]
> ..and why not add the full graph to each source detail page?

We don't know which kustomizations the source applies to, so we can't get the data to draw the graph

**Comment by stefanprodan** [comment]
I don’t understand what this does, are we displaying the Kubernetes objects reconciled by a Kustomization? If so then many things are missing, like service accounts, rbac, services, network policies, etc

File: ui/components/DependencyGraph.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  .edgePath path {
+    stroke: #333;
+    stroke-width: 1.5px;
+  }
+`;
+
+// Stolen from here:
+// https://dagrejs.github.io/project/dagre-d3/latest/demo/sentence-tokenization.html
+function DependencyGraph({ className, nodes, edges }: Props) {
```
Comment (stefanprodan) [inline ui/components/DependencyGraph.tsx:37]: Why is this called dependency graph?

**Review by stefanprodan — CHANGES_REQUESTED** [review]

**Comment by jpellizzari** [comment]
@stefanprodan 
> I don’t understand what this does, are we displaying the Kubernetes objects reconciled by a Kustomization? If so then many things are missing, like service accounts, rbac, services, network policies, etc

The PR description says "initial concept" because this will probably change drastically. I expect the topology, styling and location of this visualization to change in the near future. As of this PR, it gives us a baseline to work with to make this graph more useful and complete before release.

The intent of this type of visualization is to explain relationships between objects, build trust, and potentially provide contextual navigation.

File: ui/components/DependencyGraph.tsx (hunk 1 above)
Comment (jpellizzari) [reply ui/components/DependencyGraph.tsx]: Fixed

**Comment by stefanprodan** [comment]
@jpellizzari wouldn't be better to render the dependency graph based on `dependsOn`? Listing the objects reconciled by a Kustomization without pagination is not going to work, people have hundreds or event thousands of objects in a repo.

**Comment by jpellizzari** [comment]
> the dependency graph based on dependsOn

@stefanprodan My understanding is that would be render-able 

> @jpellizzari wouldn't be better to render the dependency graph based on `dependsOn`?

@stefanprodan My understand of the `DependsOn` field is that it would illustrate a different type of relationship between kustomizations. 

The story I am trying to tell with this visualization is "here are all of your flux objects and the resulting k8s objects that were created".

> Kustomization without pagination is not going to work, people have hundreds or event thousands of objects in a repo

Good call out; would it be viable to scope the graph down to object type and namespace? I am imagining an selector that switches between "Deployments, StatefulSet, ServiceAccount" being shown on the graph.

**Comment by stefanprodan** [comment]
@jpellizzari can you please use https://github.com/fluxcd/flux2-kustomize-helm-example I think one important thing to show in such a graph are are all the kustomizations and helmreleases that originate from the main one, keep in mind that's common to have thins like `kustomization:flux-system -> k:infra + k:apps -> hr:infra + hr:apps -> kube:workloads` and so on.

**Comment by jpellizzari** [comment]
Created a follow up issue here: https://github.com/fluxcd/webui/issues/16

**Review by stefanprodan — APPROVED** [review]
LGTM

I would changed the title to "Workloads graph" to make it obvious that HelmReleases and any other Flux CR is not displayed here.

**Pull request merged**